packages/oauth-client/src/OAuthClient.ts

```diff
@@ -0,0 +1,228 @@
+   * @param {string} params.loginUrl - see example
+   * @param {string} params.redirectUri - see example
+   *
+   * @example
+   *
+   * const oAuth2Provider = new OAuthClient({
+   *   clientId: 'my-client-id',
+   *   clientPassword: 'my-client-password',
+   *   loginUrl: 'https://earthdata.login.nasa.gov',
+   *   redirectUri: 'http://my-api.com'
+   * });
+   */
+  constructor(
+    params: {
+      clientId: string,
+      clientPassword: string,
+      loginUrl: string,
+      redirectUri: string
+    }
+  ) {
+    const {
+      clientId,
+      clientPassword,
+      loginUrl,
+      redirectUri,
+    } = params;
+
+    if (!clientId) throw new TypeError('clientId is required');
+    this.clientId = clientId;
+
+    if (!clientPassword) throw new TypeError('clientPassword is required');
+    this.clientPassword = clientPassword;
+
+    if (!loginUrl) throw new TypeError('loginUrl is required');
+    validateUrl(loginUrl);
+    this.loginUrl = loginUrl;
+
+    if (!redirectUri) throw new TypeError('redirectUri is required');
+    validateUrl(redirectUri);
+    this.redirectUri = redirectUri;
+  }
+
+  /**
+   * Get a URL of the Login authorization endpoint
+   *
+   * @param {string} [state] - an optional state to pass to login Client
+   * @returns {string} the Login authorization URL
+   */
+  getAuthorizationUrl(state?: string): string {
+    const url = new URL('/oauth/authorize', this.loginUrl);
+
+    url.searchParams.set('client_id', this.clientId);
+    url.searchParams.set('redirect_uri', this.redirectUri);
+    url.searchParams.set('response_type', 'code');
+
+    if (state) {
+      url.searchParams.set('state', state);
+    }
+
+    return url.toString();
+  }
+
+  private requestAccessToken(authorizationCode: string) {
+    return <CancelableRequest<AccessTokenResponse>>(this.postRequest({
+      loginPath: 'oauth/token',
+      form: {
+        grant_type: 'authorization_code',
+        code: authorizationCode,
+        redirect_uri: this.redirectUri,
+      },
+    }));
+  }
+
+  /**
+   * Given an authorization code, request an access token and associated
+   * information from the login service.
+   *
+   * Returns an object with the following properties:
+   *
+   * - accessToken
+   * - refreshToken
+   * - username
+   * - expirationTime (in seconds)
+   *
+   * @param {string} authorizationCode - an OAuth2 authorization code
+   * @returns {Promise<Object>} access token information
+   */
+  async getAccessToken(authorizationCode: string): Promise<Object> {
+    if (!authorizationCode) throw new TypeError('authorizationCode is required');
+    const response = await this.requestAccessToken(authorizationCode);
+
+    return {
+      accessToken: response.body.access_token,
+      refreshToken: response.body.refresh_token,
+      username: response.body.endpoint.split('/').pop(),
+      // expires_in value is in seconds
+      expirationTime: Math.floor(Date.now() / 1000) + response.body.expires_in,
+    };
+  }
```

Comment:
No error handling on the parent class. These methods should just throw and rely on the inheriting class to handle.